http-client/src/main/java/com/facebook/airlift/http/client/thrift/ThriftResponseHandler.java

```
@@ -0,0 +1,146 @@
+
+    public ThriftResponseHandler(ThriftCodec<T> thriftCodec)
+    {
+        this.thriftCodec = thriftCodec;
+    }
+
+    @Override
+    public ThriftResponse handleException(Request request, Exception exception)
+    {
+        throw ResponseHandlerUtils.propagate(request, exception);
+    }
+
+    @Override
+    public ThriftResponse handle(Request request, Response response)
+    {
+        try {
+            return new ThriftResponse(response.getStatusCode(), response.getStatusMessage(), response.getHeaders(), thriftCodec, response.getInputStream());
+        }
+        catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    public class ThriftResponse
+    {
+        private final int statusCode;
+        private final String statusMessage;
+        private final ListMultimap<HeaderName, String> headers;
+        private final T value;
+        private final IllegalArgumentException exception;
+
+        ThriftResponse(int statusCode, String statusMessage, ListMultimap<HeaderName, String> headers, ThriftCodec<T> thriftCodec, InputStream thriftBytes)
+        {
+            this.statusCode = statusCode;
+            this.statusMessage = statusMessage;
+            this.headers = ImmutableListMultimap.copyOf(headers);
+            T value = null;
+            IllegalArgumentException exception = null;
+            ThriftProtocol thriftProtocol = getThriftProtocol(headers);
+            try {
+                value = ThriftProtocolUtils.read(thriftCodec, thriftProtocol, thriftBytes);
+            }
+            catch (Exception e) {
+                exception = new IllegalArgumentException("Unable to create " + thriftCodec.getType() + " from THRIFT response", e);
+            }
+            this.value = value;
+            this.exception = exception;
+        }
+
+        private ThriftProtocol getThriftProtocol(ListMultimap<HeaderName, String> headers)
+        {
+            ThriftProtocol thriftProtocol = BINARY;
```

Let's throw an exception instead of assuming the protocol